### PR TITLE
DR-169 stairway exceptions

### PR DESCRIPTION
**Pull request by ddietterich**
- change stairway database to JSON-serialize exceptions instead of message text
- change flight status to return exception
- change JobService to throw returned exceptions on error
- add JobService method to handle flight status and either throw or return the desired class; that is now used by StudyService for the case of REST API endpoint that is synchronous and uses stairway
- have stairway catch exceptions from steps and, based on the exception, generate the proper step result. This saves us from writing repetitive try-catch in our steps.

### CODE REVIEW

File: src/main/java/bio/terra/stairway/Stairway.java [hunk 1]
```diff
@@ -140,12 +141,18 @@ public void waitForFlight(String flightId) {
 
     /**
      * Get the state of a specific flight
+     * If the flight is complete and still in our in-memory map, we remove it.
+     * The logic is that if getFlightState is called, then either the wait
```
Comment (thathert) [inline src/main/java/bio/terra/stairway/Stairway.java:145]: Missing the end of this sentence

**Review by thathert — APPROVED** [review]
Looks good to me

File: src/main/java/bio/terra/controller/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -54,7 +65,7 @@ public ErrorModel catchallHandler(Exception ex) {
         return buildErrorModel(ex);
     }
 
-    private ErrorModel buildErrorModel(Exception ex) {
+    private ErrorModel buildErrorModel(Throwable ex) {
```
Comment (ahaessly) [inline src/main/java/bio/terra/controller/GlobalExceptionHandler.java:70]: since you changed this to Throwable, should we also change our catchallHandler to handle Throwable?

File: src/main/java/bio/terra/pdao/PrimaryDataAccess.java [hunk 3]
```diff
@@ -29,7 +29,7 @@
  *     </ol>
  * </ul>
  *
- * All implementation-specific exceptions should be caught by the implementation and wrapped
+ * All implementation-specific exception should be caught by the implementation and wrapped
```
Comment (ahaessly) [inline src/main/java/bio/terra/pdao/PrimaryDataAccess.java]: seems like an unnecessary change

File: src/main/java/bio/terra/stairway/Flight.java [hunk 4]
```diff
@@ -169,13 +170,23 @@ private StepResult stepWithRetry() throws InterruptedException {
...
+                if (context().isDoing()) {
+                    result = currentStep.step.doStep(context());
+                } else {
+                    result = currentStep.step.undoStep(context());
+                }
 
+            } catch (Exception ex) {
+                // The purpose of this catch is to relieve steps of implementing their own repetitive try-catch
+                // simply to turn exceptions into StepResults.
+                StepStatus stepStatus = (ex instanceof RetryException)
```
Comment (ahaessly) [inline src/main/java/bio/terra/stairway/Flight.java:184]: Where does a RetryException actually get thrown? I don't think the individual Step should throw it. I only saw it used in a test fixture.

**Review by ahaessly — COMMENTED** [review]
couple of comments

File: src/main/java/bio/terra/controller/GlobalExceptionHandler.java (hunk 2 above)
Comment (ddietterich) [reply src/main/java/bio/terra/controller/GlobalExceptionHandler.java:70]: Sure. I changed it to Throwable, because I am pulling out the "cause" from the exception to unwrap the nesting.

File: src/main/java/bio/terra/pdao/PrimaryDataAccess.java (hunk 3 above)
Comment (ddietterich) [reply src/main/java/bio/terra/pdao/PrimaryDataAccess.java]: Yes, I did a rename and it went too far. I caught a few of these. I'll fix this one, also. Thanks

File: src/main/java/bio/terra/stairway/Flight.java (hunk 4 above)
Comment (ddietterich) [reply src/main/java/bio/terra/stairway/Flight.java:184]: My idea is that steps could throw exceptions derived from RetryException and Stairway would retry. The other way available to steps is to return StatusResult with the RETRY status.

If when we implement retrying and we don't use the exception, then I can remove it.

Comment (ahaessly) [reply src/main/java/bio/terra/stairway/Flight.java:184]: Since we aren't forcing the Step code to handle exceptions, I thought it would be the flight code that would decide to do the retry if it caught an exception.

File: src/main/java/bio/terra/controller/GlobalExceptionHandler.java [hunk 5]
```diff
@@ -46,15 +47,25 @@ public ErrorModel validationExceptionHandler(Exception ex) {
...
+    public ErrorModel jobResponseExceptionHandler(Exception ex) {
+        Throwable nestedException = ex.getCause();
+        return buildErrorModel(nestedException);
+    }
+
     // -- catchall - log so we can understand what we have missed in the handlers above
-    @ExceptionHandler(Exception.class)
+    @ExceptionHandler(Throwable.class)
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public ErrorModel catchallHandler(Exception ex) {
```
Comment (ahaessly) [inline src/main/java/bio/terra/controller/GlobalExceptionHandler.java]: I think this type now needs to be Throwable. 
I also noticed in StepResult that Throwable has been changed to Exception. I'm not sure which we really want, but I think we should be consistent here and in StepResult. 
My understanding is that Throwable are things that the application typically can not handle, like out of memory. Not sure if those are things we should be returning to the client.

File: src/main/java/bio/terra/stairway/Flight.java (hunk 4 above)
Comment (ddietterich) [reply src/main/java/bio/terra/stairway/Flight.java:184]: Without something telling it the difference, Stairway won't know which exceptions to retry and which to undo. With this PR, there are two ways we could code things:

1. For steps that will retry, they will have to implement exception handling for the retryable exceptions and return the StatusResult with RETRY status.

2. Or in the DAO and PDAO, we catch failures we want to retry and throw exceptions derived from RetryException. That let's stairway know that it should retry instead of undo-ing.

I think we will find that #2 is easier to deal with, but since we haven't done much coding for retry, I'm not sure which we will prefer.
Does that make sense?

File: src/main/java/bio/terra/controller/GlobalExceptionHandler.java (hunk 5 above)
Comment (ddietterich) [reply src/main/java/bio/terra/controller/GlobalExceptionHandler.java]: Right. I think the only thing we should deal with in our code is Exception. We never catch any non-exception. I don't think this code needs to deal with Throwables.

The only reason to have Throwable in the buildErrorModel is because the type of getCause() in the JobResponseException handler is Throwable.

If it reads better, I can make buildErrorModel take Exception and duplicate the code inside of the special case of JobResponseException.

Comment (ahaessly) [reply src/main/java/bio/terra/controller/GlobalExceptionHandler.java]: ah, maybe it is better to go back to the code in the original PR. and maybe add a comment for buildErrorModel that says it takes a Throwable because that is the type of the contained cause

**Review by ahaessly — APPROVED** [review]
👍

File: src/main/java/bio/terra/controller/GlobalExceptionHandler.java (hunk 5 above)
Comment (ddietterich) [reply src/main/java/bio/terra/controller/GlobalExceptionHandler.java]: OK, now I use Exception everywhere except that one method; and I commented that method.
Thanks for your help!
